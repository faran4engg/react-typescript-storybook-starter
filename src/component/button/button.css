.button {
    padding:.5rem 1rem;
    margin: 0 1rem;
    font-family: inherit;
    border:1px solid #999999;
    color: #ffffff;
}

.button-primary {
    background-color:#7B68EE;

}
.button-primary:hover {
    background-color: #4835bb;
}

.button-secondary {
    background-color: #3CB371;
}
.button-secondary:hover {
    background-color: #098040;
}

.button-ghost {
    background-color: transparent;
    border:1px solid currentColor;
    color: #7B68EE;
}

.button-text {
    background-color:transparent;
    color: inherit;
    border:0;
}

