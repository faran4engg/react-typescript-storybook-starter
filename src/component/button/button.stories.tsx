import React from "react";
import { Button } from "./button";
import { text, select } from "@storybook/addon-knobs";
import { action } from "@storybook/addon-actions";

export default {
  title: "Components/Button",
  component: Button
};

export const Usage = () => (
  <Button
    onClick={action("Button Clicked")}
    as={select("Show As", ["primary", "secondary", "ghost", "text"], "primary")}
    type={select("Type", ["button", "submit"], "button")}
  >
    {text("Text", "Primary Button")}
  </Button>
);
