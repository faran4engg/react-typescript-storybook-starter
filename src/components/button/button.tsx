import React, { MouseEvent } from "react";

import "./button.css";

export interface ButtonProps {
  /**optionally add an ID to the element */
  id?: string;
  /** Button Type: submit or button */
  type?: "submit" | "button";
  /** Button Content: Subcomponents or text */
  children: React.ReactNode;
  /** Display As: primary, secondary, ghost, or plain text (for additional styling) */
  as: "primary" | "secondary" | "ghost" | "text";
  /** Function to run when the button is clicked */
  onClick: (event: MouseEvent) => void;
  /** Button is disabled: true or false */
  disabled?: boolean;
  /** Additional CSS classes to apply */
  className?: string;
}

export const Button: React.FC<ButtonProps> = ({
  type = "button",
  children,
  as = "primary",
  onClick,
  disabled = false,
  className = "",
  id = ""
}: ButtonProps) => {
  const classes = `button button-${as} ${className}`.trim();
  return (
    <>
      <button
        id={id}
        type={type}
        onClick={onClick}
        className={classes}
        disabled={disabled}
      >
        {children}
      </button>
    </>
  );
};
