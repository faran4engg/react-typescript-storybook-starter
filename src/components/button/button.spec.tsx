import React from "react";
import { cleanup, render, fireEvent } from "@testing-library/react";

import { Button } from "./button";

afterEach(cleanup);

describe("Button ", () => {
  it("should match snapshot", () => {
    const testFn = jest.fn();
    const el = render(
      <Button type="button" onClick={testFn} as="primary">
        hello world
      </Button>
    );
    expect(el.baseElement).toMatchSnapshot();
  });

  it("should render", () => {
    const testFn = jest.fn();
    const { getByText } = render(
      <Button type="button" onClick={testFn} as="primary">
        hello world
      </Button>
    );
    getByText(/hello world/i); //if found returns true
  });

  it("should render as a primary button", () => {
    const testFn = jest.fn();
    const { getByRole } = render(
      <Button type="button" onClick={testFn} as="primary">
        hello world
      </Button>
    );
    const el = getByRole("button");
    expect(el.classList.contains("button-primary")).toBe(true);
  });

  it("should render as a secondary button", () => {
    const testFn = jest.fn();
    const { getByRole } = render(
      <Button type="button" onClick={testFn} as="secondary">
        hello world
      </Button>
    );
    const el = getByRole("button");
    expect(el.classList.contains("button-secondary")).toBe(true);
  });

  it("should render as a ghost button", () => {
    const testFn = jest.fn();
    const { getByRole } = render(
      <Button type="button" onClick={testFn} as="ghost">
        hello world
      </Button>
    );
    const el = getByRole("button");
    expect(el.classList.contains("button-ghost")).toBe(true);
  });

  it("should render as a text button", () => {
    const testFn = jest.fn();
    const { getByRole } = render(
      <Button type="button" onClick={testFn} as="text">
        hello world
      </Button>
    );
    const el = getByRole("button");
    expect(el.classList.contains("button-text")).toBe(true);
  });

  it("Spy should be called when clicked", () => {
    const testFn = jest.fn();
    const { getByText } = render(
      <Button type="button" onClick={testFn} as="primary">
        hello world
      </Button>
    );
    const btn = getByText(/hello world/i);
    fireEvent.click(btn);
    expect(testFn).toHaveBeenCalledTimes(1);
    fireEvent.click(btn);
    expect(testFn).toHaveBeenCalledTimes(2);
  });

  it("should be disabled", () => {
    const testFn = jest.fn();
    const { getByRole } = render(
      <Button type="button" onClick={testFn} as="primary" disabled>
        hello world
      </Button>
    );
    const el = getByRole("button");
    //null would mean the attribute doesnt exist empty string means it exists
    expect(el.getAttribute("disabled")).toBe("");
  });
});
