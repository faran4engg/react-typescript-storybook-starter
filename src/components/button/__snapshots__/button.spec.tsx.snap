// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button  should match snapshot 1`] = `
<body>
  <div>
    <button
      class="button button-primary"
      id=""
      type="button"
    >
      hello world
    </button>
  </div>
</body>
`;
